### ISSUE TITLE
Tcprewrite: UDP checksum calculated wrong for fragmented UDP packet

### ISSUE THREAD

**Issue by sorenhede** [appneta/tcpreplay#190]
When using -C on pcap-file containing fragmented UDP packets, the UDP checksum is incorrect.

It is possible to reproduce the issue using the file "http://loki.stockton.edu/~olanm/csis4222/homework/fragments.pcap" 
With:
tcprewrite -C --infile=fragments.pcap --outfile=output.pcap

View of the fragments.pcap file in Wireshark
![fragments](https://cloud.githubusercontent.com/assets/10185732/8332106/5a441a82-1a8c-11e5-922d-0879365930d6.png)

View of the output.pcap with incorrect checksum after tcprewrite -C in Wireshark
![output](https://cloud.githubusercontent.com/assets/10185732/8332111/5e210872-1a8c-11e5-8f2d-02e893173329.png)

My current tcprewite version:
tcprewrite version: 4.1.0 (build git:v4.1.0)
Copyright 2013-2014 by Fred Klassen <tcpreplay at appneta dot com> - AppNeta Inc.
Copyright 2000-2012 by Aaron Turner <aturner at synfin dot net>
The entire Tcpreplay Suite is licensed under the GPLv3
Cache file supported: 04
Not compiled with libdnet.
Compiled against libpcap: 1.5.3
64 bit packet counters: enabled
Verbose printing via tcpdump: enabled
Fragroute engine: disabled

**Comment by fklassen** [comment]
Hi, any chance that you have UDP checksum offload enabled on the NIC? 

I'll schedule some time to look into this for next beta release.

**Comment by sorenhede** [comment]
Hi, the fragments.pcap testfile used in this example have correct UDP checksum - but when pass it thrugh tcprewrite with -C the UDP checksum become incorrect in the output file.

I believe there might be a problem with calculation of fragmented UDP checksum

**Comment by fklassen** [comment]
It appears that #225 fixes this bug, mostly.

The underlying issue is that that `tcpreplay-edit` and `tcprewrite` can only process one packet at a time. For many operations, if something is edited, packet checksums must be updated. Original technique was to recalculate the entire checksum. This doesn't work for fragments, which require checksumming frames that will be sent/rewritten in future frames.

Fix includes implementing [RFC 1624 Incremental Checksum](https://tools.ietf.org/html/rfc1624). It combines local edit differences and then applies it to the already calculated checksum.

This means that if edits are done to only the first frame of a fragmented packet, then everything works. Those options include...
```
  -r, --portmap=str          Rewrite TCP/UDP ports
   -s, --seed=num             Randomize src/dst IPv4/v6 addresses w/ given seed
   -N, --pnat=str             Rewrite IPv4/v6 addresses using pseudo-NAT
   -S, --srcipmap=str         Rewrite source IPv4/v6 addresses using pseudo-NAT
   -D, --dstipmap=str         Rewrite destination IPv4/v6 addresses using pseudo-NAT
   -e, --endpoints=str        Rewrite IP addresses to be between two endpoints
   -b, --skipbroadcast        Skip rewriting broadcast/multicast IPv4/v6 addresses
   -C, --fixcsum              Force recalculation of IPv4/TCP/UDP header checksums
   -E, --efcs                 Remove Ethernet checksums (FCS) from end of frames
       --ttl=str              Modify the IPv4/v6 TTL/Hop Limit
       --tos=num              Set the IPv4 TOS/DiffServ/ECN byte
       --tclass=num           Set the IPv6 Traffic Class byte
       --flowlabel=num        Set the IPv6 Flow Label
```

The following options require processing of follow-on packets. Options affected are...
```
   -m, --mtu=num              Override default MTU length (1500 bytes)
       --mtu-trunc            Truncate packets larger then specified MTU
```
In these cases, we need special handling when the first frame of a fragmented packet is processed. For IPv4 UDP, the checksum is set to `0`, which disables checksum. However this trick doesn't work for IPv6 UDP, which requires checksums for every packet. Also doesn't work for TCP, however [RFC 1191 PMTU Discovery](https://tools.ietf.org/html/rfc1191) make fragmented TCP packets a near impossibility on networks.

**Comment by thelaughingbandit** [comment]
I am using version 4.1.1 of tcprewrite and I am still having this issue. The captured packets have a correct udp checksum in wireshark, but when I use the --portmap option to change the source, destination or both, the checksums are no longer valid for those udp packets that are fragmented.

When I ran ./configure, I didn't use any options. Is there something in particular I need to get this functionality working?

Many thanks in advance,

Andrew

**Comment by fklassen** [comment]
Just make sure you are using this source code instead of cloning from github https://github.com/appneta/tcpreplay/releases

**Comment by thelaughingbandit** [comment]
Thanks very much for the quick response. I built everything again from scratch from your link but I had the same issue. In the end I managed to get around it by using --portmap for a single port at a time. When I try to port map a range of ports to a single port, the udp checksums are incorrect.

**Comment by fklassen** [comment]
OK, sounds like a bug. I've got it scheduled for next release.

**Comment by fklassen** [comment]
Fixed in #294. UDP checksum updates are to be bypassed for fragments. Corrected the logic to do so.
